linera-indexer/src/indexer.rs

```diff
@@ -0,0 +1,244 @@
+struct IndexerConfig {
+    /// The port of the indexer server
+    #[structopt(long, default_value = "8081")]
+    port: u16,
+    /// The port of the service to connect
+    #[structopt(long = "service-port", default_value = "8080")]
+    service_port: u16,
+    /// The address of the service to connect
+    #[structopt(long = "service-address", default_value = "localhost")]
+    service_address: String,
+    /// TLS/SSl for service connexion
+    #[structopt(long = "tls")]
+    tls: bool,
+    /// Rocksdb storage path
+    #[structopt(long, default_value = "./indexer.db")]
+    storage: String,
+    /// The maximal number of simultaneous stream queries to the database
+    #[structopt(long, default_value = "10")]
+    max_stream_queries: usize,
+    /// Cashe size for the rocksdb storage
+    #[structopt(long, default_value = "1000")]
+    cache_size: usize,
+    /// Height to start the indexing
+    #[structopt(long, default_value = "0")]
+    start: BlockHeight,
+    /// Indexer command
+    #[structopt(subcommand)]
+    command: IndexerCommand,
+}
+
+struct TokioSpawner(tokio::runtime::Handle);
+impl futures::task::Spawn for TokioSpawner {
+    fn spawn_obj(
+        &self,
+        obj: futures::task::FutureObj<'static, ()>,
+    ) -> Result<(), futures::task::SpawnError> {
+        self.0.spawn(obj);
+        Ok(())
+    }
+}
+
+enum Protocol {
+    Http,
+    Websocket,
+}
+
+type Context = RocksDbContext<()>;
+
+fn service_address(config: &IndexerConfig, protocol: Protocol) -> String {
+    let protocol = match protocol {
+        Protocol::Http => "http",
+        Protocol::Websocket => "ws",
+    };
+    let tls = if config.tls { "s" } else { "" };
+    format!(
+        "{}{}://{}:{}",
+        protocol, tls, config.service_address, config.service_port
+    )
+}
+
+/// Connects to the websocket of the service node for a particular chain
+async fn connect(
+    config: &IndexerConfig,
+    indexer: &Indexer<Context>,
+    chain_id: ChainId,
+) -> Result<ChainId, IndexerError> {
+    let mut request =
+        format!("{}/ws", service_address(config, Protocol::Websocket)).into_client_request()?;
+    request.headers_mut().insert(
+        "Sec-WebSocket-Protocol",
+        HeaderValue::from_str("graphql-transport-ws")?,
+    );
+    let (connection, _) = async_tungstenite::tokio::connect_async(request).await?;
+    let (sink, stream) = connection.split();
+
+    let mut client = GraphQLClientClientBuilder::new()
+        .build(
+            stream,
+            sink,
+            TokioSpawner(tokio::runtime::Handle::current()),
+        )
+        .await?;
+    let operation: StreamingOperation<Notifications> =
+        graphql_ws_client::graphql::StreamingOperation::new(notifications::Variables { chain_id });
+
+    let mut stream = client.streaming_operation(operation).await?;
+
+    while let Some(item) = stream.next().await {
+        if let Ok(response) = item {
+            if let Some(data) = response.data {
```

Comment:
Should we print the error if not?

Reply:
Probably, I was hesitating to close the WebSocket in case of `Err`.
For the null data, it is not supposed to happen with this webSocket so it is safe to just print an error as you say.

Reply:
Yes, we can print an error and not close the WebSocket.